lib/reporting/mfa_report.rb

```diff
@@ -0,0 +1,218 @@
+
+      def self.all_events
+        constants.map { |c| const_get(c) }
+      end
+    end
+
+    module Methods
+      def self.all_methods
+        TwoFactorAuthenticatable::AuthMethod.constants.map do |c|
+          next if c == :PHISHING_RESISTANT_METHODS || c == :REMEMBER_DEVICE
+
+          if c == :PERSONAL_KEY
+            # The AuthMethod constant defines this as `personal_key`
+            # but in the events it is `personal-key`
+            'personal-key'
+          else
+            TwoFactorAuthenticatable::AuthMethod.const_get(c)
+          end
+        end.compact
+      end
+    end
+
+    # @param [Array<String>] issuers
+    # @param [Range<Time>] time_range
+    def initialize(
+      issuers:,
+      time_range:,
+      verbose: false,
+      progress: false,
+      slice: 1.day,
+      threads: 10
+    )
+      @issuers = issuers
+      @time_range = time_range
+      @verbose = verbose
+      @progress = progress
+      @slice = slice
+      @threads = threads
+    end
+
+    def verbose?
+      @verbose
+    end
+
+    def progress?
+      @progress
+    end
+
+    def as_tables
+      [
+        overview_table,
+        multi_factor_auth_table,
+      ]
+    end
+
+    def as_emailable_reports
+      [
+        Reporting::EmailableReport.new(
+          title: 'Overview',
+          table: overview_table,
+        ),
+        Reporting::EmailableReport.new(
+          title: 'Multi Factor Authentication Metrics',
+          table: multi_factor_auth_table,
+        ),
+      ]
+    end
+
+    def to_csvs
+      as_tables.map do |table|
+        CSV.generate do |csv|
+          table.each do |row|
+            csv << row
+          end
+        end
+      end
+    end
+
+    # @return Array<Hash>
+    def data
+      @data ||= begin
+        fetch_results
+      end
+    end
+
+    def fetch_results
+      cloudwatch_client.fetch(query:, from: time_range.begin, to: time_range.end)
+    end
+
+    def query
+      params = {
+        issuers: quote(issuers),
+        event_names: quote(Events.all_events),
+        fields:,
+        stats:,
+      }
+
+      format(<<~QUERY, params)
+        fields %{fields}
+        | filter properties.service_provider IN %{issuers}
+        | filter name in %{event_names} and not properties.event_properties.confirmation_for_add_phone and properties.event_properties.context != 'reauthentication'
+        | filter properties.event_properties.success = '1'
+        | stats %{stats}
+      QUERY
+    end
+
+    def fields
+      Methods.all_methods.map do |method|
+        # Cloudwatch doesn't like the hyphen in personal-key
+        no_hypen_method = method.tr('-', '_')
+        'properties.event_properties.multi_factor_auth_method = ' +
+          "#{method}' as #{no_hypen_method}"
+      end.join(', ')
+    end
+
+    def stats
+      Methods.all_methods.map do |method|
+        m = method.tr('-', '_')
+        "sum(#{m}) as `#{m + '_total'}`"
+      end.join(', ')
+    end
```

Comment:
I like the idea of dynamically generating these, but seeing as there's so many exceptions (skipping "remember device" above, renaming personal key, etc), WDYT of just having one big string where the entire query is written out? might be easier to debug?

Reply:
i don't love either way, so happy to make it a big string instead!

Reply:
updated in https://github.com/18F/identity-idp/pull/10076/commits/690ebf9fa0667b9e101e9de60df90a53a1eefda0